File: site/src/main/java/cn/edu/zjnu/acm/judge/sandbox/win32/Job.java
package cn.edu.zjnu.acm.judge.sandbox.win32;

import java.io.Closeable;
import jnc.platform.win32.JOBOBJECTINFOCLASS;
import jnc.platform.win32.JOBOBJECT_BASIC_LIMIT_INFORMATION;
import jnc.platform.win32.JOBOBJECT_BASIC_UI_RESTRICTIONS;
import jnc.platform.win32.JOBOBJECT_INFORMATION;
import jnc.platform.win32.Kernel32;
import jnc.platform.win32.Kernel32Util;

import static jnc.platform.win32.JOBOBJECTINFOCLASS.JobObjectBasicLimitInformation;
import static jnc.platform.win32.JOBOBJECTINFOCLASS.JobObjectBasicUIRestrictions;
import static jnc.platform.win32.WinNT.JOB_OBJECT_LIMIT_ACTIVE_PROCESS;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_DESKTOP;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_DISPLAYSETTINGS;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_EXITWINDOWS;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_GLOBALATOMS;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_HANDLES;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_READCLIPBOARD;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_SYSTEMPARAMETERS;
import static jnc.platform.win32.WinNT.JOB_OBJECT_UILIMIT_WRITECLIPBOARD;

/**
 * @author zhanhb
 */
public class Job implements Closeable {

    private final long /*HANDLE*/ hJob;

    public Job() {
        long handle = Kernel32.INSTANCE.CreateJobObjectW(null, null);
        Kernel32Util.assertTrue(handle != 0);
        this.hJob = handle;
    }

    private void setInformationJobObject(JOBOBJECTINFOCLASS jobobjectinfoclass, JOBOBJECT_INFORMATION jobj) {
        Kernel32Util.assertTrue(Kernel32.INSTANCE.SetInformationJobObject(hJob, jobobjectinfoclass.value(), jobj, jobj.size()));
    }

    public void init() {
        JOBOBJECT_BASIC_LIMIT_INFORMATION jobli = new JOBOBJECT_BASIC_LIMIT_INFORMATION();
        jobli.setActiveProcessLimit(1);
        // These are the only 1 restrictions I want placed on the job (process).
        jobli.setLimitFlags(JOB_OBJECT_LIMIT_ACTIVE_PROCESS);
        setInformationJobObject(JobObjectBasicLimitInformation, jobli);

        // Second, set some UI restrictions.
        JOBOBJECT_BASIC_UI_RESTRICTIONS jbur = new JOBOBJECT_BASIC_UI_RESTRICTIONS();
        jbur.setUiRestrictionsClass(
                // The process can't access USER objects (such as other windows)
                // in the system.
                JOB_OBJECT_UILIMIT_HANDLES
                | JOB_OBJECT_UILIMIT_READCLIPBOARD
                | JOB_OBJECT_UILIMIT_WRITECLIPBOARD
                | JOB_OBJECT_UILIMIT_SYSTEMPARAMETERS
                | JOB_OBJECT_UILIMIT_DISPLAYSETTINGS
                | JOB_OBJECT_UILIMIT_GLOBALATOMS
                | // Prevents processes associated with the job from creating desktops
                // and switching desktops using the CreateDesktop and SwitchDesktop functions.
                JOB_OBJECT_UILIMIT_DESKTOP
                | // The process can't log off the system.
                JOB_OBJECT_UILIMIT_EXITWINDOWS);
        setInformationJobObject(JobObjectBasicUIRestrictions, jbur);
    }

    public void assignProcess(long /*HANDLE*/ hProcess) {
        Kernel32Util.assertTrue(Kernel32.INSTANCE.AssignProcessToJobObject(hJob, hProcess));
    }

    @Override
    public void close() {
        Handle.close(hJob);
    }

}

File: site/src/main/kotlin/cn/edu/zjnu/acm/judge/sandbox/win32/Job.kt
package cn.edu.zjnu.acm.judge.sandbox.win32

import jnc.platform.win32.*
import jnc.platform.win32.JOBOBJECTINFOCLASS.JobObjectBasicLimitInformation
import jnc.platform.win32.JOBOBJECTINFOCLASS.JobObjectBasicUIRestrictions
import jnc.platform.win32.WinNT.*
import java.io.Closeable

class Job : Closeable {

    private val /*HANDLE*/ hJob: Long

    init {
        val handle = Kernel32.INSTANCE.CreateJobObjectW(null, null)
        Kernel32Util.assertTrue(handle != 0L)
        this.hJob = handle
    }

    private fun setInformationJobObject(jobobjectinfoclass: JOBOBJECTINFOCLASS, jobj: JOBOBJECT_INFORMATION) {
        Kernel32Util.assertTrue(Kernel32.INSTANCE.SetInformationJobObject(hJob, jobobjectinfoclass.value(), jobj, jobj.size()))
    }

    fun init() {
        val jobli = JOBOBJECT_BASIC_LIMIT_INFORMATION()
        jobli.activeProcessLimit =1
        // These are the only 1 restrictions I want placed on the job (process).
        jobli.limitFlags = JOB_OBJECT_LIMIT_ACTIVE_PROCESS
        setInformationJobObject(JobObjectBasicLimitInformation, jobli)

        // Second, set some UI restrictions.
        val jbur = JOBOBJECT_BASIC_UI_RESTRICTIONS()
        jbur.uiRestrictionsClass = (
                // The process can't access USER objects (such as other windows)
                // in the system.
                JOB_OBJECT_UILIMIT_HANDLES
                        or JOB_OBJECT_UILIMIT_READCLIPBOARD
                        or JOB_OBJECT_UILIMIT_WRITECLIPBOARD
                        or JOB_OBJECT_UILIMIT_SYSTEMPARAMETERS
                        or JOB_OBJECT_UILIMIT_DISPLAYSETTINGS
                        or JOB_OBJECT_UILIMIT_GLOBALATOMS
                        or // Prevents processes associated with the job from creating desktops
                        // and switching desktops using the CreateDesktop and SwitchDesktop functions.
                        JOB_OBJECT_UILIMIT_DESKTOP
                        or // The process can't log off the system.
                        JOB_OBJECT_UILIMIT_EXITWINDOWS)
        setInformationJobObject(JobObjectBasicUIRestrictions, jbur)
    }

    fun assignProcess(/*HANDLE*/ hProcess: Long) {
        Kernel32Util.assertTrue(Kernel32.INSTANCE.AssignProcessToJobObject(hJob, hProcess))
    }

    override fun close() {
        Handle.close(hJob)
    }

}
